### PR TITLE
[ACR] Add validations in donwload_blob() and get_manifest()

### PR DESCRIPTION
**Pull request by YalinLi0312**
Resolves https://github.com/Azure/azure-sdk-for-python/issues/30039
- In donwload_blob(): 
  - validate the `Content-Range` value in response header while getting blob size.
  - only read up to its length.
- In get_manifest():
  - also validate the digest calculated from content with the `Docker-Content-Digest` in response header, we would catch malicious or corrupted content anyway.
  - add manifest size check to protect from malicious unbounded responses. The valid manifest should be smaller than 4MB.

### CODE REVIEW

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 1]
```diff
@@ -942,21 +945,21 @@ def get_manifest(self, repository: str, tag_or_digest: str, **kwargs) -> GetMani
...
                 raise DigestValidationError(
                     "The content of retrieved manifest digest does not match the requested digest."
                 )
-        else:
-            digest = response.http_response.headers['Docker-Content-Digest']
-            if not _validate_digest(manifest_bytes, digest):
-                raise DigestValidationError(
-                    "The server-computed digest does not match the client-computed digest."
-                )
+        digest = response.http_response.headers['Docker-Content-Digest']
```
Comment (xiangyan99) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:958]: Now you changed the logic: you will check this even tag_or_digest.startswith("sha256:").

Is this by design?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 2]
```diff
@@ -942,21 +945,21 @@ def get_manifest(self, repository: str, tag_or_digest: str, **kwargs) -> GetMani
                 **kwargs
             )
         )
+        if response.http_response.headers['Content-Length'] > MAX_MANIFEST_SIZE:
+            raise ValueError("Manifest size is bigger than 4MB.")
```
Comment (kristapratico) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: nit: Not sure if we standardized the error message across languages already, but I wonder if we should rephrase this to "Manifest size must be smaller than 4MB" so it's clear that >4MB is too big

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_download_stream.py [hunk 3]
```diff
@@ -68,7 +68,7 @@ async def __anext__(self) -> bytes:
         try:
             return await self._yield_data()
         except StopAsyncIteration:
-            if self._downloaded >= self._blob_size:
+            if self._downloaded == self._blob_size:
```
Comment (kristapratico) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_download_stream.py:71]: do we already have test cases that cover this?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 2 above)
Comment (xiangyan99) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: What was the error before? Is this a breaking change?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 1 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:958]: Now we changed to validate with the digest in response headers when digest is passing.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 2 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: I think it should be fine, it's consistent with other languages.

Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: It's not a breaking change, we didn't check the size before. It should be able to download successfully before.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_download_stream.py [hunk 4]
```diff
@@ -68,7 +68,7 @@ def __next__(self) -> bytes:
         try:
             return self._yield_data()
         except StopIteration:
-            if self._downloaded >= self._blob_size:
+            if self._downloaded == self._blob_size:
```
Comment (xiangyan99) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_download_stream.py:71]: what if self._downloaded > self._blob_size?

Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_download_stream.py:71]: - self._downloaded should always `<=` self.blob_size. 
- downloaded is possible larger than blob_size when downloading a chunk more than one time, but we don't support retry when downloading blob.
So I think `>` should not happen.

Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_download_stream.py]: Revert the change, it will keep downloading when downloaded data is larger than blob_size in some cases.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_download_stream.py (hunk 3 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/aio/_async_download_stream.py]: Reverted this update. Yes, we have. There's digest validation tests, it will check by digest when it stops downloading.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 2 above)
Comment (kristapratico) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: Just checking my understanding - without the added validation in the PR, this would download successfully before? Why are we adding the validation?

Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: Yes, it was downloading successfully. We changed to add manifest size validation following [contained](https://github.com/containerd/containerd/blob/main/remotes/docker/resolver.go#L59) motivation and to protect from malicious unbounded responses.

Comment (kristapratico) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: Ah okay, so this is a true GA breaking change made for security reasons. Did this get approved by architects already?

Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: It's a decision made across languages, no apiview updates with this updates, I don't know if Anna has approved. cc: @annatisch to help.

Comment (annatisch) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: If this change is being made across languages - there should be an email with the archboard about this?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 1 above)
Comment (annatisch) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:958]: So we're validating _twice_ as default behaviour? That seem excessive?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 2 above)
Comment (annelo-msft) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: This is consistent with behavior of tooling in the ecosystem, e.g. containerd.  https://github.com/Azure/azure-sdk-for-net/issues/35771

Comment (lmolkova) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:949]: > Is it possible to have a malicious manifest < 4MB?

Yes, but the goal is not to protect from malicious content, but to fail fast. Small manifests are fine, but if they are too big,  we won't spend resources in client applications attempting to download, buffer and calculate digest of invalid data a.

Comment (annatisch) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py]: Thanks for the additional context!
If there's no valid manifests >4MB that would have previously been working then this should be fine :)

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py [hunk 5]
```diff
@@ -942,21 +947,24 @@ def get_manifest(self, repository: str, tag_or_digest: str, **kwargs) -> GetMani
                 **kwargs
             )
         )
+        manifest_size = _get_manifest_size(response.http_response.headers)
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:950]: So the content-length header will always represent the _total_ manifest size? Not just the chunk?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 1 above)
Comment (annatisch) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:963]: As discussed offline - validating the data twice is fine, but we should need to compute the digest twice as this is the time consuming part and doesn't need to be duplicated.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_helpers.py [hunk 6]
```diff
@@ -139,3 +140,19 @@ def _compute_digest(data: Union[IO[bytes], bytes]) -> str:
 def _validate_digest(data: Union[IO[bytes], bytes], expected_digest: str) -> bool:
     digest = _compute_digest(data)
     return digest == expected_digest
+
+def _get_blob_size(headers: Dict[str, str]) -> int:
+    if not headers["Content-Range"]:
+        raise ValueError("Missing content-range header in response.")
+    blob_size = int(headers["Content-Range"].split("/")[1])
+    if blob_size <= 0:
+        raise ValueError("Invalid content-range header in response.")
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_helpers.py:149]: Could we please add the value the Content-Range header to the error message so it's really easy to see why it might be invalid without debugging?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_helpers.py [hunk 7]
```diff
@@ -139,3 +140,19 @@ def _compute_digest(data: Union[IO[bytes], bytes]) -> str:
...
+    return blob_size
+
+def _get_manifest_size(headers: Dict[str, str]) -> int:
+    if not headers["Content-Length"]:
+        raise ValueError("Missing content-length header in response.")
+    manifest_size = int(headers["Content-Length"])
+    if manifest_size <= 0:
+        raise ValueError("Invalid content-length header in response.")
```
Comment (annatisch) [inline sdk/containerregistry/azure-containerregistry/azure/containerregistry/_helpers.py:157]: Can we please add the value of the invalid header to the error message?

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 5 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:950]: Yes. - manifest is not downloading by chunks, it is downloading in one time. So the content-length header represent the content size in this download call, which is the total manifest size.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 1 above)
Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:963]: Yeah, compute digest is wrapped inside _validate_digest(), I'll think about just comparing with two digests.

Comment (YalinLi0312) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:964]: Changed to compare digests here.

File: sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py (hunk 5 above)
Comment (annatisch) [reply sdk/containerregistry/azure-containerregistry/azure/containerregistry/_container_registry_client.py:950]: Ah yes - I was confusing it with the blob download :P

**Review by annatisch — APPROVED** [review]
